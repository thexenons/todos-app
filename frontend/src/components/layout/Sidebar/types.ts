import { PropsWithChildren } from "react";

export type SidebarProps = PropsWithChildren;
