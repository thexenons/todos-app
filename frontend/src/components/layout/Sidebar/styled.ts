import styled, { css } from "styled-components";

import Button from "../../atoms/Button";

export const SidebarButton = styled(Button)`
	align-self: flex-end;
	${({ theme }) => theme.mixins.margin({ all: 0.5 })}
`;

export const SidebarButtonContent = styled.span`
	transform: rotate(90deg);
	transition: transform 0.3s;

	${({ theme }) => theme.media.md.up} {
		transform: rotate(0deg);
	}
`;

export const SidebarContentWrapper = styled.div`
	overflow: hidden;
	flex: 1 1 auto;
	transition: height 0.3s, width 0.3s;
`;

export const SidebarContent = styled.div`
	${({ theme }) => theme.mixins.padding({ all: 2 })}
	${({ theme }) =>
		theme.mixins.flex({ direction: "column", gap: theme.spacing(2) })}
`;

export const SidebarWrapper = styled.aside<{
	$isOpen: boolean;
	$height: number;
}>`
	position: relative;
	${({ theme }) => theme.mixins.flex({ direction: "column" })}
	background-color: ${({ theme }) => theme.colors.background.paper};
	color: ${({ theme }) => theme.colors.text.primary};

	${({ $isOpen, $height }) =>
		$isOpen
			? css`
					${SidebarContentWrapper} {
						height: ${$height ? `${$height}px` : "auto"};
					}

					${SidebarButtonContent} {
						transform: rotate(-90deg);
					}
			  `
			: css`
					${SidebarContentWrapper} {
						height: 0px;
					}
			  `}

	${({ theme }) => theme.media.md.up} {
		${({ $isOpen }) =>
			$isOpen
				? css`
						${SidebarContentWrapper} {
							width: 240px;
						}

						${SidebarButtonContent} {
							transform: rotate(180deg);
						}
				  `
				: css`
						${SidebarContentWrapper} {
							width: 0px;
						}
				  `}
	}
`;
