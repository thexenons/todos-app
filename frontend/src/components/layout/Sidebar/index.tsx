import { useAtom } from "jotai";
import { FC, useCallback } from "react";
import { useElementSize } from "usehooks-ts";

import { sidebarIsOpenAtom } from "../../../state/ui";
import { ButtonVariants } from "../../atoms/Button/types";
import * as S from "./styled";
import type { SidebarProps } from "./types";

const Sidebar: FC<SidebarProps> = ({
	children,
	isSidebarOpen: isSidebarOpenProps,
	setSidebarOpen: setSidebarOpenProps,
}) => {
	const [isSidebarOpen, setSidebarOpen] = useAtom(sidebarIsOpenAtom);
	const finalIsSidebarOpen = isSidebarOpenProps ?? isSidebarOpen;
	const finalSetSidebarOpen = setSidebarOpenProps ?? setSidebarOpen;
	const [setNode, sizes] = useElementSize();

	const toggleSidebar = useCallback(
		() => finalSetSidebarOpen?.(!finalIsSidebarOpen),
		[finalIsSidebarOpen, finalSetSidebarOpen]
	);

	return (
		<S.SidebarWrapper $isOpen={!!finalIsSidebarOpen} $height={sizes.height}>
			<S.SidebarButton
				isRounded
				variant={ButtonVariants.outlined}
				onClick={toggleSidebar}
			>
				<S.SidebarButtonContent>{">"}</S.SidebarButtonContent>
			</S.SidebarButton>
			<S.SidebarContentWrapper>
				<S.SidebarContent ref={setNode}>{children}</S.SidebarContent>
			</S.SidebarContentWrapper>
		</S.SidebarWrapper>
	);
};

export default Sidebar;
