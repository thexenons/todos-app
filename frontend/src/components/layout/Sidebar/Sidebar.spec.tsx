import { render } from "@testing-library/react";

import BaseTestComponent from "../../../tests/BaseTestComponent";
import Sidebar from ".";

describe("Sidebar", () => {
	it("Render", () => {
		const { container } = render(
			<BaseTestComponent>
				<Sidebar />
			</BaseTestComponent>
		);
		expect(container).toMatchSnapshot();
	});
});
