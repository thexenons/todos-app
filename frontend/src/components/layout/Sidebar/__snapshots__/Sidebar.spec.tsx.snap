// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar > Render 1`] = `
<div>
  <aside
    class="styled__SidebarWrapper-sc-6c82ddb3-4 kgzCTB"
  >
    <button
      class="styled__ButtonWrapper-sc-f8cd241f-0 dlIeAl styled__SidebarButton-sc-6c82ddb3-0 jegNMZ"
    >
      <svg
        class="styled__SidebarButtonContent-sc-6c82ddb3-1 kInmfU"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
        />
      </svg>
    </button>
    <div
      class="styled__SidebarContentWrapper-sc-6c82ddb3-2 kgXZjq"
    >
      <div
        class="styled__SidebarContent-sc-6c82ddb3-3 jSXxdi"
      />
    </div>
  </aside>
</div>
`;
